exchanges/kraken/kraken_websocket.go

```diff
@@ -0,0 +1,676 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// WsUnsubscribeToChannel sends a request to WS to unsubscribe to supplied channel name and pairs
+func (k *Kraken) WsUnsubscribeToChannel(topic string, currencies []string, requestID int64) error {
+	resp := WebsocketSubscriptionEventRequest{
+		Event: krakenWsUnsubscribe,
+		Pairs: currencies,
+		Subscription: WebsocketSubscriptionData{
+			Name: topic,
+		},
+	}
+	if requestID > 0 {
+		resp.RequestID = requestID
+	}
+	json, err := common.JSONEncode(resp)
+	if err != nil {
+		return err
+	}
+	err = k.writeToWebsocket(string(json))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// WsUnsubscribeToChannelByChannelID sends a request to WS to unsubscribe to supplied channel ID
+func (k *Kraken) WsUnsubscribeToChannelByChannelID(channelID int64) error {
+	resp := WebsocketUnsubscribeByChannelIDEventRequest{
+		Event:     krakenWsUnsubscribe,
+		ChannelID: channelID,
+	}
+	json, err := common.JSONEncode(resp)
+	if err != nil {
+		return err
+	}
+	err = k.writeToWebsocket(string(json))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// addNewSubscriptionChannelData stores channel ids, pairs and subscription types to an array
+// allowing correlation between subscriptions and returned data
+func addNewSubscriptionChannelData(response *WebsocketEventResponse) {
+	for i := range subscriptionChannelPair {
+		if response.ChannelID == subscriptionChannelPair[i].ChannelID {
+			return
+		}
+	}
+
+	// We change the / to - to maintain compatibility with REST/config
+	pair := currency.NewPairWithDelimiter(response.Pair.Base.String(), response.Pair.Quote.String(), "-")
+	subscriptionChannelPair = append(subscriptionChannelPair, WebsocketChannelData{
+		Subscription: response.Subscription.Name,
+		Pair:         pair,
+		ChannelID:    response.ChannelID,
+	})
+}
+
+// getSubscriptionChannelData retrieves WebsocketChannelData based on response ID
+func getSubscriptionChannelData(id float64) WebsocketChannelData {
+	for i := range subscriptionChannelPair {
+		if id == subscriptionChannelPair[i].ChannelID {
+			return subscriptionChannelPair[i]
+		}
+	}
+	return WebsocketChannelData{}
+}
+
+// resubscribeToChannel will attempt to unsubscribe and resubscribe to a channel
+func (k *Kraken) ResubscribeToChannel(channel string, pair currency.Pair) {
+	// Kraken WS formats pairs with / but config and REST use -
+	formattedPair := strings.ToUpper(strings.Replace(pair.String(), "-", "/", 1))
+	if krakenWsResubscribeFailureLimit > 0 {
+		var successfulUnsubscribe bool
+		for i := 0; i < krakenWsResubscribeFailureLimit; i++ {
+			err := k.WsUnsubscribeToChannel(channel, []string{formattedPair}, 0)
+			if err != nil {
+				log.Error(err)
+				time.Sleep(krakenWsResubscribeDelayInSeconds * time.Second)
+				continue
+			}
+			successfulUnsubscribe = true
+			break
+		}
+		if !successfulUnsubscribe {
+			log.Fatalf("%v websocket channel %v failed to unsubscribe after %v attempts",
```

Comment:
This fatal might be better as something else? We don't want the bot to fail due to this sub system.

Reply:
[Like with the last PR , this is more an interim solution](https://github.com/thrasher-/gocryptotrader/pull/252#discussion_r266289513). If gct cant successfully subscribe after that many attempts then there is something severely wrong. This will be addressed more thoroughly in future

Reply:
ok good to know!